### ISSUE TITLE
Support for batch-based attacks?

### ISSUE THREAD

**Issue by iamgroot42** [tensorflow/cleverhans#1110]
***The issue tracker should only be used to report bugs or feature requests. If you are looking for support from other library users, please ask a question on [StackOverflow](https://stackoverflow.com/questions/tagged/cleverhans).***

**Is your feature request related to a problem? Please describe.**
Just like most machine-learning based libraries, users should have the option to perform batched attacks. Some attacks may not support parallelism that well, but we can at least offer the option for them to send in all data with a batch size, instead of them batching it themselves and then sending it (some attacks like MadryEtAl cannot be run altogether on a lot of data)

**Describe the solution you'd like**
A simple loop to iterate through data in batches (resting in the generic Attack class, maybe)? Uses the batch_size attack_parameter (make it standard), uses that batch size to iterate through data batch-wise. We could also check for actual batched support for attacks (for Eg, MadryEtAl doesn't support currently in terms of flags) and pass that on while performing the batched attacks

**Describe alternatives you've considered**
As a user, I just break my data into batches and then concentrate all batch-wise perturbed data batches together. It's not too long a workaround, but it would definitely be favorable to have such a functionality behind the curtains in the library itself.

**Additional context**
I could work on this solution, if the maintainers think this could be useful. Should not take a lot of time (just some basic changes in the attack module, along with maybe some tests?)

**Comment by npapernot** [comment]
Most attacks should be able to take a batch of data as the first dimension of the input tensor. Can you provide details of attacks and framework version where you find this to be an issue?

**Comment by iamgroot42** [comment]
Yes, but the batch size used by the attack algorithm does not seem to be a controllable parameter. For instance, running MadryEtAl on a large set of data fails since it tries to compute everything in one go. Maybe support for a 'batch_size' param could help call this attack directly, instead of users breaking data into chunks and then combining data themselves in the end?

**Comment by npapernot** [comment]
Can you provide details of attacks and framework version where you find this to be an issue?

**Comment by iamgroot42** [comment]
I tried to generate adversarial examples on the test set of CIFAR-10 together in one go for a Resnet-50 model, for which I got an internal OOM error. It would be nice to have a 'batch_size' parameter (if I am not wrong, it is already present for some attacks eg HopSkipJumpAttack) for this attack. We could have a basic batch-breaking operation in the base class, and functions which already implement batching could use their batching instead. 

Cleverhans version:  3.0.1
Tensorflow version:  1.14.0
